src/serac/physics/utilities/functional/domain_integral.hpp

```diff
@@ -197,7 +201,48 @@ void gradient_kernel(const mfem::Vector& dU, mfem::Vector& dR, derivatives_type*
   }
 }
 
-}  // namespace domain_integral
+template < typename T1, typename T2 >
+struct linear_approximation {
+  T1 value;
+  T2 derivative;
+};
+
+template < typename element_type, int dim >
+auto evaluate_shape_functions(tensor < double, dim > xi, tensor< double, dim, dim > J) {
+  if constexpr (element_type::family == Family::HCURL) {
+    auto N = dot(element_type::shape_functions(xi), inv(J));
+    auto curl_N = element_type::shape_function_curl(xi) / det(J);
+    if constexpr (dim == 3) {
+      curl_N = dot(curl_N, transpose(J));
+    }
+
+    using pair_t = linear_approximation< 
+      typename std::remove_reference<decltype(N[0])>::type, 
+      typename std::remove_reference<decltype(curl_N[0])>::type
+    >;
+    tensor < pair_t, element_type::ndof > output{};
+    for (int i = 0; i < element_type::ndof; i++) {
+      output[i].value = N[i];
+      output[i].derivative = curl_N[i];
+    }
+    return output;
+  } else {
+    auto N = element_type::shape_functions(xi);
+    auto grad_N = dot(element_type::shape_function_gradients(xi), inv(J));
+
+    using pair_t = linear_approximation< 
```

Comment:
At first glance, it is difficult to understand what this does. I think this method could use some comments.